README.md
<p align="center">
  <a href="https://flutter.dev">
    <img src="https://img.shields.io/badge/Platform-Flutter-02569B?logo=flutter"
      alt="Platform" />
  </a>
  <a href="https://pub.dartlang.org/packages/multi_border">
    <img src="https://img.shields.io/pub/v/multi_border" alt="Pub Package" />
  </a>
  <a href="https://opensource.org/licenses/MIT">
    <img src="https://img.shields.io/github/license/benjamin-otto/multi_border" alt="License"/>
  </a>
</p>

<p>  
  <img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/packageTitle.png?raw=true" alt="MultiBorder Package" width="100%"/>
</p>

<p align="center">
A custom decoration for quickly adding multiple borders of varying colors and widths to your widgets.
</p>

## Features

- Set each border's color.
- Set each border's width.
- Set a border radius for rectangular borders.
- Rectangular or circular borders.
- `MultiBorderDecoration.evenBorders()` constructor simplifies creating borders of even widths.

## Getting started

Import:

```dart
import 'package:multi_border/multi_border.dart';
```

Add a `MultiBorderDecoration` to your `Container` widget:

```dart
Container(
  decoration: MultiBorderDecoration(
    borderSides: [
      BorderSide(color: Colors.blue.shade100, width: 8),
      BorderSide(color: Colors.blue.shade200, width: 12),
      BorderSide(color: Colors.blue.shade300, width: 16)
    ],
  ),
  child: const Text('Wrap me with colorful borders'),
);
```

Or, use the `MultiBorderDecoration.evenBorders()` constructor to create borders of equal width.
- This will create a border for each color in [borderColors] of width [width].

```dart
Container(
  decoration: MultiBorderDecoration.evenBorders(
    width: 8,
    borderColors: [
      Colors.blue.shade100,
      Colors.blue.shade200,
      Colors.blue.shade300,
    ],
  ),
  child: const Text('Wrap me with colorful borders'),
);
```

## Examples

### Sharp rectangle with even borders:

<img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/evenBordersSharpRectangle.png?raw=true" alt="Even Borders Sharp Rectangle" width="150"/>

```dart
Container(
  decoration: MultiBorderDecoration.evenBorders(
    width: 8,
    borderColors: const [
      Color(0xFFF14A42),
      Color(0xFFF27D3B),
      Color(0xFFF2A83A),
      Color(0xFFF8C43F),
    ],
  ),
  child: Container(
    width: 100,
    height: 100,
    color: const Color(0xFF4B3D2F),
    child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
  ),
);
```

### Rounded rectangle with even borders:

<img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/evenBordersRoundedRectangle.png?raw=true" alt="Rounded Rectangle Even Borders" width="150"/>

```dart
Container(
  clipBehavior: Clip.hardEdge,
  decoration: MultiBorderDecoration.evenBorders(
    width: 8,
    innerRadius: BorderRadius.circular(8),
    borderColors: const [
      Color(0xFFF14A42),
      Color(0xFFF27D3B),
      Color(0xFFF2A83A),
      Color(0xFFF8C43F),
    ],
  ),
  child: Container(
    width: 100,
    height: 100,
    color: const Color(0xFF4B3D2F),
    child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
  ),
);
```

### Rounded rectangle with varying width borders:

<img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/customBordersRectangle.png?raw=true" alt="Rectangle Varying Borders" width="150"/>

```dart
Container(
  clipBehavior: Clip.hardEdge,
  decoration: MultiBorderDecoration(
    innerRadius: const BorderRadius.only(
      topLeft: Radius.circular(16),
      bottomRight: Radius.circular(16),
    ),
    borderSides: const [
      BorderSide(color: Color(0xFFF14A42), width: 4),
      BorderSide(color: Color(0xFFF27D3B), width: 8),
      BorderSide(color: Color(0xFFF2A83A), width: 12),
      BorderSide(color: Color(0xFFF8C43F), width: 16),
    ],
  ),
  child: Container(
    width: 100,
    height: 100,
    color: const Color(0xFF4B3D2F),
    child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
  ),
);
```

### Circle with even borders:

<img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/evenBordersCircle.png?raw=true" alt="Circle Even Borders" width="150"/>

```dart
Container(
  clipBehavior: Clip.antiAlias,
  decoration: MultiBorderDecoration.evenBorders(
    shape: BoxShape.circle,
    width: 8,
    borderColors: const [
      Color(0xFFF14A42),
      Color(0xFFF27D3B),
      Color(0xFFF2A83A),
      Color(0xFFF8C43F),
    ],
  ),
  child: Container(
    width: 100,
    height: 100,
    color: const Color(0xFF4B3D2F),
    child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
  ),
);
```



### Circle with varying width borders:

<img src="https://github.com/benjamin-otto/multi_border/blob/main/screenshots/customBordersCircle.png?raw=true" alt="Circle Varying Borders" width="150"/>

```dart
Container(
  clipBehavior: Clip.antiAlias,
  decoration: MultiBorderDecoration(
    shape: BoxShape.circle,
    borderSides: const [
      BorderSide(color: Color(0xFFF14A42), width: 4),
      BorderSide(color: Color(0xFFF27D3B), width: 8),
      BorderSide(color: Color(0xFFF2A83A), width: 12),
      BorderSide(color: Color(0xFFF8C43F), width: 16),
    ],
  ),
  child: Container(
    width: 100,
    height: 100,
    color: const Color(0xFF4B3D2F),
    child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
  ),
);
```

CHANGELOG.md
## 1.0.0

### Breaking changes

* Refactor to draw borders by extending ShapeDecoration instead of using a CustomPainter.

## 0.0.2

* Add a more detailed package description.

## 0.0.1

* Initial release.


lib/src/multi_border.dart
import 'package:flutter/material.dart';

import 'extensions.dart';

/// A decoration class for adding multiple borders to widgets.
class MultiBorderDecoration extends ShapeDecoration {
  /// Creates a border for each BorderSide supplied in [borderSides].
  MultiBorderDecoration({
    required this.borderSides,
    this.boxShape = BoxShape.rectangle,
    this.innerRadius = BorderRadius.zero,
    super.color,
    super.gradient,
    super.image,
    super.shadows,
  })  : assert(borderSides.isNotEmpty),
        assert(
            boxShape == BoxShape.rectangle || innerRadius == BorderRadius.zero,
            '[boxShape] set to BoxShape.circle does not require an [innerRadius]'),
        super(
          shape: (boxShape == BoxShape.rectangle)
              ? _compositeRRectBorder(borderSides, innerRadius)
              : _compositeCircleBorder(borderSides),
        );

  /// Creates a border for each Color supplied in [borderColors].
  ///
  /// Each border will be of equal width [width].
  factory MultiBorderDecoration.evenBorders({
    required List<Color> borderColors,
    BoxShape boxShape = BoxShape.rectangle,
    BorderRadius innerRadius = BorderRadius.zero,
    double width = 1.0,
    Color? color,
    Gradient? gradient,
    DecorationImage? image,
    List<BoxShadow>? shadows,
  }) {
    final borderSides = borderColors.fold(
      <BorderSide>[],
      (borderSides, borderColor) =>
          borderSides..add(BorderSide(color: borderColor, width: width)),
    );

    return MultiBorderDecoration(
      borderSides: borderSides,
      boxShape: boxShape,
      innerRadius: innerRadius,
      color: color,
      gradient: gradient,
      image: image,
      shadows: shadows,
    );
  }

  /// Describes the [color] and [width] of each border.
  final List<BorderSide> borderSides;

  /// [BoxShape.rectangle] or [BoxShape.circle]
  final BoxShape boxShape;

  /// The radius of the innermost border to be drawn.
  final BorderRadius innerRadius;

  static ShapeBorder _compositeRRectBorder(
    List<BorderSide> borderSides,
    BorderRadius innerRadius,
  ) {
    final shapeBorders = <ShapeBorder>[];
    double radiusIncrement = 0.0;

    for (int i = 0; i < borderSides.length; i++) {
      shapeBorders.add(
        RoundedRectangleBorder(
          side: borderSides[i],
          borderRadius: innerRadius.incrementNonZeroCorners(radiusIncrement),
        ),
      );
      if (i + 1 < borderSides.length) {
        final nextDiff = borderSides[i + 1].width - borderSides[i].width;
        radiusIncrement += borderSides[i].width + nextDiff;
      }
    }

    return shapeBorders.reduce((tot, next) => tot + next);
  }

  static ShapeBorder _compositeCircleBorder(List<BorderSide> borderSides) {
    return borderSides.sublist(1).fold(
          CircleBorder(side: borderSides[0], eccentricity: 1.0),
          (prevCircleBorder, borderSide) =>
              prevCircleBorder +
              CircleBorder(side: borderSide, eccentricity: 1.0),
        );
  }

  @override
  Path getClipPath(Rect rect, TextDirection textDirection) {
    return shape.getInnerPath(rect, textDirection: textDirection);
  }
}


lib/src/multi_border_painter.dart
import 'package:flutter/widgets.dart';

import 'mb_border.dart';
import 'extensions.dart';

/// Draws the [MultiBorder] borders.
class MultiBorderPainter extends CustomPainter {
  MultiBorderPainter({
    required this.shape,
    required this.borders,
    required this.innerRadius,
  });

  /// The shape of the border.
  ///
  /// [BoxShape.rectangle] or [BoxShape.circle]
  final BoxShape shape;

  /// The color and width of each border.
  final List<MBBorder> borders;

  /// The innermost radius that will be added upon for each more outward border.
  final BorderRadius innerRadius;

  @override
  void paint(Canvas canvas, Size size) {
    switch (shape) {
      case BoxShape.rectangle:
        _drawRectangleBorders(canvas, size);
        break;
      case BoxShape.circle:
        _drawCircleBorders(canvas, size);
        break;
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => false;

  void _drawRectangleBorders(Canvas canvas, Size size) {
    double totalWidth = borders.totalWidth;
    double widthToPaint = borders.totalWidth;

    for (MBBorder border in borders) {
      final halfBorderWidth = border.width / 2;
      final offset = widthToPaint - halfBorderWidth;
      final painted = totalWidth - widthToPaint;
      final additionalRadius = painted + halfBorderWidth;

      final rrect = RRect.fromLTRBAndCorners(
        offset,
        offset,
        size.width - offset,
        size.height - offset,
        topLeft: innerRadius.topLeft.accrueRounded(additionalRadius),
        topRight: innerRadius.topRight.accrueRounded(additionalRadius),
        bottomLeft: innerRadius.bottomLeft.accrueRounded(additionalRadius),
        bottomRight: innerRadius.bottomRight.accrueRounded(additionalRadius),
      );

      final paint = Paint()
        ..color = border.color
        ..strokeWidth = border.width
        ..style = PaintingStyle.stroke;

      canvas.drawRRect(rrect, paint);
      widthToPaint -= border.width;
    }
  }

  void _drawCircleBorders(Canvas canvas, Size size) {
    double widthToPaint = borders.totalWidth;

    for (MBBorder border in borders) {
      final paint = Paint()
        ..color = border.color
        ..strokeWidth = border.width
        ..style = PaintingStyle.stroke;

      canvas.drawCircle(
        Offset(size.width / 2, size.height / 2),
        (size.shortestSide / 2) + (border.width / 2) - widthToPaint,
        paint,
      );

      widthToPaint -= border.width;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_border/multi_border.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'MultiBorderDecoration Examples',
      home: Scaffold(
        body: Center(child: Examples()),
      ),
    );
  }
}

class Examples extends StatelessWidget {
  const Examples({super.key});

  @override
  Widget build(BuildContext context) {
    return ListView.separated(
      itemCount: _examples.length,
      separatorBuilder: (_, __) => const SizedBox(height: 24),
      itemBuilder: (_, i) => Center(child: _examples[i]),
    );
  }

  List<Widget> get _examples => [
        _rectangleEvenBorders,
        _roundedRectangleEvenBorders,
        _roundedRectangleVaryingBorders,
        _circleVaryingBorders,
        _circleEvenBorders,
      ];

  List<Color> get _borderColors => const [
        Color(0xFFF14A42),
        Color(0xFFF27D3B),
        Color(0xFFF2A83A),
        Color(0xFFF8C43F),
      ];

  List<BorderSide> get _borderSides => const [
        BorderSide(color: Color(0xFFF14A42), width: 4),
        BorderSide(color: Color(0xFFF27D3B), width: 8),
        BorderSide(color: Color(0xFFF2A83A), width: 12),
        BorderSide(color: Color(0xFFF8C43F), width: 16),
      ];

  Container get _rectangleEvenBorders => Container(
        decoration: MultiBorderDecoration.evenBorders(
          width: 8,
          borderColors: _borderColors,
        ),
        child: _child,
      );

  Container get _roundedRectangleEvenBorders => Container(
        clipBehavior: Clip.hardEdge,
        decoration: MultiBorderDecoration.evenBorders(
          width: 8,
          innerRadius: BorderRadius.circular(8),
          borderColors: _borderColors,
        ),
        child: _child,
      );

  Container get _roundedRectangleVaryingBorders => Container(
        clipBehavior: Clip.hardEdge,
        decoration: MultiBorderDecoration(
          borderSides: _borderSides,
          innerRadius: const BorderRadius.only(
            topLeft: Radius.circular(16),
            bottomRight: Radius.circular(16),
          ),
        ),
        child: _child,
      );

  Container get _circleEvenBorders => Container(
        clipBehavior: Clip.antiAlias,
        decoration: MultiBorderDecoration.evenBorders(
          boxShape: BoxShape.circle,
          width: 8,
          borderColors: _borderColors,
        ),
        child: _child,
      );

  Container get _circleVaryingBorders => Container(
        clipBehavior: Clip.antiAlias,
        decoration: MultiBorderDecoration(
          boxShape: BoxShape.circle,
          borderSides: _borderSides,
        ),
        child: _child,
      );

  Container get _child => Container(
        width: 100,
        height: 100,
        color: const Color(0xFF4B3D2F),
        child: const Icon(Icons.thumb_up, color: Color(0xFFCEC1B6)),
      );
}
